### PR TITLE
IF: Merge main into hotstuff_integration

### LINKED ISSUE #1509: IF: Update BLS integration to use HotStuff version from main
The current commit of the `hotstuff_integration` branch (https://github.com/AntelopeIO/leap/commits/005d500e7b9a29325be0fd253710a5fe205f1792) uses an old version of BLS code that has since been modified as part of the effort of merging the code into the `main` branch of Leap.

The `hotstuff_integration` branch should be changed to fix the discrepancy of BLS between the branch and `main`. Also additional immediate merge conflicts should be resolved.

### PR DESCRIPTION
**Pull request by heifner**
Included in this merge:
- Move to new `bls12-381` library and away from `bls-signature` library. `bls_private_key`, `bls_public_key`, `bls_signature` types have been reimplemented with `bls12-381` `g1`, `g2` types.
- Fix for memory leak in `bls_private_key::generate()`
  -- Is there a strict aliasing issue with this implementation?
- Cleanup of `bls_private_key` implementation.
- Removed bls_* types from `abi_serializer`
- Fixed `libtester` by not including `bls12-381` header files in `controller.hpp` which is used by `libtester`.
- Removed `aggregate_signature` protocol feature and host functions as bls primitives are being provided to contracts instead.
- Added needed virtual destructor to `base_pacemaker`
- Moved `chain_pacemaker` from `producer_plugin` to `chain_plugin`
- Removed unneeded default constructors from hotstuff message structs.

Created issue: https://github.com/AntelopeIO/leap/issues/1531 for failing `bls_sig_verif_string_multi` test.
Created issue: https://github.com/AntelopeIO/leap/issues/1533 for failing deepmind tests.
All other tests are passing.

Resolves #1509

### CODE REVIEW

**Review by fcecin — APPROVED** [review]

File: plugins/chain_plugin/chain_plugin.cpp [hunk 1]
```diff
@@ -2640,25 +2640,26 @@ read_only::get_consensus_parameters(const get_consensus_parameters_params&, cons
...
-      results.v_height               = fs.v_height;
-      results.high_qc                = fs.high_qc;
-      results.current_qc             = fs.current_qc;
-      results.schedule               = fs.schedule;
-      for (auto proposal: fs.proposals) {
-         chain::hs_proposal_message & p = proposal.second;
-         results.proposals.push_back( hs_complete_proposal_message( p ) );
-      }
-   }
+   // TODO: move to producer_plugin
```
Comment (linh2931) [inline plugins/chain_plugin/chain_plugin.cpp:2643]: For all those TDDOs, should we use compile WARNING for now so we won't forget about them?

Comment (heifner) [reply plugins/chain_plugin/chain_plugin.cpp:2643]: This particular one, I should fix as part of this PR I think. But for others, yes.

File: libraries/chain/abi_serializer.cpp [hunk 2]
```diff
@@ -546,14 +548,15 @@ namespace eosio { namespace chain {
...
+               bool present = vo.contains(string(field.name).c_str());
+               if( present || is_optional(field.type) ) {
                   if( disallow_additional_fields )
                      EOS_THROW( pack_exception, "Unexpected field '${f}' found in input object while processing struct '${p}'",
                                 ("f", ctx.maybe_shorten(field.name))("p", ctx.get_path_string()) );
                   {
                      auto h1 = ctx.push_to_path( impl::field_path_item{ .parent_struct_itr = s_itr, .field_ordinal = i } );
                      auto h2 = ctx.disallow_extensions_unless( &field == &st.fields.back() );
-                     _variant_to_binary(_remove_bin_extension(field.type), vo[field.name], ds, ctx);
+                     _variant_to_binary(_remove_bin_extension(field.type), present ? vo[field.name] : fc::variant(nullptr), ds, ctx);
```
Comment (ericpassmore) [inline libraries/chain/abi_serializer.cpp:559]: good use of conditional operator

File: .github/actions/parallel-ctest-containers/dist/index.mjs
Comment (ericpassmore) [inline .github/actions/parallel-ctest-containers/dist/index.mjs:1]: did not review

Comment (heifner) [reply .github/actions/parallel-ctest-containers/dist/index.mjs:1]: You seem to be reviewing `main` changes. It is more important to review the merge commits.

File: libraries/chain/include/eosio/chain/hotstuff.hpp [hunk 3]
```diff
@@ -40,8 +40,6 @@ namespace eosio { namespace chain {
       fc::sha256                          proposal_id = NULL_PROPOSAL_ID; //vote on proposal
       name                                finalizer;
       fc::crypto::blslib::bls_signature   sig;
-
-      hs_vote_message() = default;
```
Comment (ericpassmore) [inline libraries/chain/include/eosio/chain/hotstuff.hpp:44]: Should we remove the other defaults
- extended_schedule() = default;
- quorum_certificate() = default;
- hs_proposal_message() = default;
- hs_new_view_message() = default;
- hs_new_block_message() = default;
- hs_proposal_message() = default;
- finalizer_state() = default;

File: libraries/chain/include/eosio/chain/symbol.hpp [hunk 4]
```diff
@@ -65,20 +65,7 @@ namespace eosio {
...
-                  auto comma_pos = s.find(',');
-                  EOS_ASSERT(comma_pos != string::npos, symbol_type_exception, "missing comma in symbol");
-                  auto prec_part = s.substr(0, comma_pos);
-                  uint8_t p = fc::to_int64(prec_part);
-                  string name_part = s.substr(comma_pos + 1);
-                  EOS_ASSERT( p <= max_precision, symbol_type_exception, "precision ${p} should be <= 18", ("p", p));
-                  return symbol(string_to_symbol(p, name_part.c_str()));
-               } FC_CAPTURE_LOG_AND_RETHROW((from))
-            }
+            static symbol from_string(const string& from);
```
Comment (ericpassmore) [inline libraries/chain/include/eosio/chain/symbol.hpp:68]: note: from PR https://github.com/AntelopeIO/leap/pull/1234

File: libraries/chain/include/eosio/chain/thread_utils.hpp [hunk 5]
```diff
@@ -37,19 +37,39 @@ namespace eosio { namespace chain {
...
+
+         std::promise<void> start_complete;
+         std::atomic<uint32_t> threads_remaining = num_threads;
+         std::exception_ptr pending_exception;
+         std::mutex pending_exception_mutex;
+
+         try {
+            for( size_t i = 0; i < num_threads; ++i ) {
+               _thread_pool.emplace_back( std::thread( &named_thread_pool::run_thread, this, i, on_except, init, std::ref(start_complete),
+                                                       std::ref(threads_remaining), std::ref(pending_exception), std::ref(pending_exception_mutex) ) );
```
Comment (ericpassmore) [inline libraries/chain/include/eosio/chain/thread_utils.hpp:63]: nitpick: 3 lines?

File: unittests/test_contracts.hpp.in [hunk 6]
```diff
@@ -10,10 +10,10 @@
       fc::read_file_contents("${CMAKE_BINARY_DIR}/unittests/" #D "/" #C "/" #C ".wasm", s); \
       return std::vector<uint8_t>(s.begin(), s.end());                                      \
    }                                                                                        \
-   static std::vector<char> CN ## _abi() {                                                  \
+   static std::string CN ## _abi() {                                                        \
       std::string s;                                                                        \
       fc::read_file_contents("${CMAKE_BINARY_DIR}/unittests/" #D "/" #C "/" #C ".abi", s);  \
-      return std::vector<char>(s.begin(), s.end());                                         \
+      return s;                                                                             \
```
Comment (ericpassmore) [inline unittests/test_contracts.hpp.in:16]: spacing?

File: libraries/chain/platform_timer_kqueue.cpp
Comment (ericpassmore) [inline libraries/chain/platform_timer_kqueue.cpp:1]: better name 👍

File: plugins/chain_plugin/chain_plugin.cpp [hunk 7]
```diff
@@ -70,10 +70,12 @@ void validate(boost::any& v,
   // one string, it's an error, and exception will be thrown.
   std::string const& s = validators::get_single_string(values);
 
- if ( s == "head" ) {
+  if ( s == "head" ) {
```
Comment (ericpassmore) [inline plugins/chain_plugin/chain_plugin.cpp:73]: nitpick: trim leading space

File: plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp [hunk 8]
```diff
@@ -700,7 +785,36 @@ class read_only {
...
+            if (p.show_payer) {
+               result.rows.emplace_back(fc::mutable_variant_object("data", std::move(data_var))("payer", row.second));
+            } else {
+               result.rows.emplace_back(std::move(data_var));
+            }            
+         }
+         result.more = p.more;
+         result.next_key = p.next_key;
+         return result;
+      };
```
Comment (ericpassmore) [inline plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp:817]: note: new addition to main

File: libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp [hunk 9]
```diff
@@ -277,11 +277,7 @@ inline constexpr auto get_intrinsic_table() {
       "env.bls_pairing",
       "env.bls_g1_map",
       "env.bls_g2_map",
-      "env.bls_fp_mod",
-      "env.bls_verify",
```
Comment (linh2931) [inline libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp:271]: At the first look, I thought `bls_verify` was removed by accident as it did not have `aggregate` as a part of its name. But you were correct.

File: plugins/chain_plugin/chain_plugin.cpp (hunk 1 above)
Comment (heifner) [reply plugins/chain_plugin/chain_plugin.cpp]: Moved implementation from `producer_plugin` to `chain_plugin`.

File: libraries/chain/include/eosio/chain/hotstuff.hpp (hunk 3 above)
Comment (heifner) [reply libraries/chain/include/eosio/chain/hotstuff.hpp:44]: removed

**Review by ericpassmore — APPROVED** [review]

File: libraries/chain/controller.cpp [hunk 10]
```diff
@@ -27,7 +27,7 @@
 #include <eosio/chain/platform_timer.hpp>
 #include <eosio/chain/deep_mind.hpp>
 #include <eosio/chain/wasm_interface_collection.hpp>
-#include <eosio/chain/hotstuff.hpp>
+//#include <eosio/chain/hotstuff.hpp>
```
Comment (linh2931) [inline libraries/chain/controller.cpp:30]: This can be removed next round.

**Review by linh2931 — APPROVED** [review]

**Comment by greg7mdp** [comment]
Just noticed these g++ warnings:
```
[3/307] Building CXX object libraries/libfc/CMakeFiles/fc.dir/src/crypto/bls_public_key.cpp.o
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_public_key.cpp: In function ‘bls12_381::g1 fc::crypto::blslib::parse_base58(const std::string&)’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_public_key.cpp:34:22: warning: unused variable ‘prefix’ [-Wunused-variable]
   34 |       constexpr auto prefix = config::bls_public_key_base_prefix;
      |                      ^~~~~~
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_public_key.cpp: In member function ‘bool fc::crypto::blslib::bls_public_key::valid() const’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_public_key.cpp:57:4: warning: no return statement in function returning non-void [-Wreturn-type]
   57 |    }
      |    ^
[4/307] Building CXX object libraries/libfc/CMakeFiles/fc.dir/src/crypto/bls_private_key.cpp.o
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_private_key.cpp: In function ‘std::vector<unsigned char> fc::crypto::blslib::priv_parse_base58(const std::string&)’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_private_key.cpp:45:18: warning: unused variable ‘pivot’ [-Wunused-variable]
   45 |       const auto pivot = base58str.find('_');
      |                  ^~~~~
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_private_key.cpp:60:4: warning: no return statement in function returning non-void [-Wreturn-type]
   60 |    }
      |    ^
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_private_key.cpp: In member function ‘std::string fc::crypto::blslib::bls_private_key::to_string(const fc::yield_function_t&) const’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_private_key.cpp:75:4: warning: no return statement in function returning non-void [-Wreturn-type]
   75 |    }
      |    ^
[5/307] Building CXX object libraries/libfc/CMakeFiles/fc.dir/src/crypto/bls_signature.cpp.o
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp: In member function ‘std::size_t fc::crypto::blslib::bls_signature::variable_size() const’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp:60:4: warning: no return statement in function returning non-void [-Wreturn-type]
   60 |    }
      |    ^
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp: In function ‘std::size_t fc::crypto::blslib::hash_value(const bls_signature&)’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp:96:4: warning: no return statement in function returning non-void [-Wreturn-type]
   96 |    }
      |    ^
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp: In function ‘bool fc::crypto::blslib::operator==(const bls_signature&, const bls_signature&)’:
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp:80:9: warning: infinite recursion detected [-Winfinite-recursion]
   80 |    bool operator == ( const bls_signature& p1, const bls_signature& p2) {
      |         ^~~~~~~~
/home/greg/github/enf/leap/libraries/libfc/src/crypto/bls_signature.cpp:81:22: note: recursive call
   81 |       return p1._sig == p2._sig;
      |              ~~~~~~~~^~~~~~~~~~
```

**Comment by heifner** [comment]
> Just noticed these g++-12 warnings, but I guess this is not relevant to this PR.

There are a number of clang warnings as well. I assume those will be handled by: https://github.com/AntelopeIO/leap/issues/1520

File: libraries/chain/controller.cpp (hunk 10 above)
Comment (heifner) [reply libraries/chain/controller.cpp]: Removed

File: libraries/libfc/include/fc/crypto/base64.hpp [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
 #include <string>
+#include <string_view>
 
 namespace fc {
 std::string base64_encode(unsigned char const* bytes_to_encode, unsigned int in_len);
 inline std::string base64_encode(char const* bytes_to_encode, unsigned int in_len) { return base64_encode( (unsigned char const*)bytes_to_encode, in_len); }
-std::string base64_encode( const std::string& enc );
-std::string base64_decode( const std::string& encoded_string);
+std::string base64_encode( const std::string_view& enc );
+std::string base64_decode( const std::string_view& encoded_string);
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/crypto/base64.hpp:9]: detail: no need to pass `std::string_view` by reference

File: libraries/libfc/include/fc/crypto/base64.hpp [hunk 12]
```diff
@@ -1,14 +1,15 @@
...
 
 std::string base64url_encode(unsigned char const* bytes_to_encode, unsigned int in_len);
 inline std::string base64url_encode(char const* bytes_to_encode, unsigned int in_len) { return base64url_encode( (unsigned char const*)bytes_to_encode, in_len); }
-std::string base64url_encode( const std::string& enc );
-std::string base64url_decode( const std::string& encoded_string);
+std::string base64url_encode( const std::string_view& enc );
+std::string base64url_decode( const std::string_view& encoded_string);
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/crypto/base64.hpp:14]: detail: no need to pass `std::string_view` by reference

File: libraries/libfc/include/fc/crypto/private_key.hpp [hunk 13]
```diff
@@ -46,14 +46,14 @@ namespace fc { namespace crypto {
...
-         explicit private_key(const string& base58str);
+         explicit private_key(const std::string& base58str);
          std::string to_string(const fc::yield_function_t& yield = fc::yield_function_t()) const;
 
       private:
          storage_type _storage;
 
          private_key( storage_type&& other_storage )
-         :_storage(forward<storage_type>(other_storage))
+            :_storage(other_storage)
```
Comment (greg7mdp) [inline libraries/libfc/include/fc/crypto/private_key.hpp:56]:
```suggestion
            :_storage(std::move(other_storage))
```

**Review by greg7mdp — APPROVED** [review]

File: libraries/libfc/include/fc/crypto/base64.hpp (hunk 11 above)
Comment (heifner) [reply libraries/libfc/include/fc/crypto/base64.hpp]: Fixed

File: libraries/libfc/include/fc/crypto/base64.hpp (hunk 12 above)
Comment (heifner) [reply libraries/libfc/include/fc/crypto/base64.hpp]: Fixed

File: libraries/libfc/include/fc/crypto/private_key.hpp (hunk 13 above)
Comment (heifner) [reply libraries/libfc/include/fc/crypto/private_key.hpp]: Fixed

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: MERGE
summary: merge main into hotstuff branch.
Note:end

**Pull request merged**